arkouda/categorical.py

```diff
@@ -529,3 +534,197 @@ def concatenate(self, others : List[Categorical], ordered : bool=True) -> Catego
+        self.name = None  # Clear our internal Categorical object name
+
+    def is_registered(self) -> np.bool_:
+        """
+         Return True iff the object is contained in the registry
+
+        Returns
+        -------
+        numpy.bool
+            Indicates if the object is contained in the registry
+
+        Raises
+        ------
+        RegistrationError
+            Raised if there's a server-side error or a mis-match of registered pieces
+
+        See Also
+        --------
+        register, attach, unregister, unregister_categorical_by_name
+
+        Notes
+        -----
+        Objects registered with the server are immune to deletion until
+        they are unregistered.
+        """
+        reg:bool
+        components = self._get_vars_with_register()
+        if components:
+            registered = [k for k, v in components if v.is_registered()]
+            if registered and len(registered) == len(components):  # everything is registered
+                reg = True
+            elif not registered:  # nothing is registered
+                reg = False
+            else:  # Some are registered but not all
+                raise RegistrationError(f"Not all registerable components of Categorical {self.name} are registered.")
+        else:
+            reg = False
+
+        return np.bool_(reg)  # Using numpy for consistency with other implementations
+
+    @staticmethod
+    @typechecked
+    def attach(user_defined_name:str) -> Categorical:
+        """
+        Function to return a Categorical object attached to the registered name in the
+        arkouda server which was registered using register()
+
+        Parameters
+        ----------
+        user_defined_name : str
+            user defined name which Categorical object was registered under
+
+        Returns
+        -------
+        Categorical
+               The Categorical object created by re-attaching to the corresponding server components
+
+       Raises
+       ------
+       TypeError
+            if user_defined_name is not a string
+
+        See Also
+        --------
+        register, is_registered, unregister, unregister_categorical_by_name
+        """
+        parts = {}  # Build dict of registered components by invoking their corresponding Class.attach functions
+        for k,v in Categorical.RegisterableComponents.items():
+            parts[k] = v.attach(f"{user_defined_name}.{k}")
+        # Now call constructor with kwargs and hand it the unpacked dictionary object
+        c = Categorical(None, **parts)
+        c.name = user_defined_name # Update our name
+        return c
+
+    @staticmethod
+    @typechecked
+    def unregister_categorical_by_name(user_defined_name:str) -> None:
+        """
+        Function to unregister Categorical object by name which was registered
+        with the arkouda server via register()
+        WARNING: This will cause the underlying object to go out-of-scope issuing a deletion
```

Comment:
This is a weird issue because I'm doing an attach which makes a new object  It's bound to the arkouda_server with a string name.  It invokes unregister (make it deletable on the arkouda_server) and then goes out of scope when the function returns.  That triggers a __del__() operation on this specific instance and it tells the arkouda_server to delete the server object associated with the name.  This means any other objects bound to it are now pointing to something which no longer exists on the server.

@mhmerrill  and I had a discussion on what to do about this PR  and I'm in the process of refactoring it, so the scope issue is now moot.

Reply:
👍